sql/insert.go

```diff
@@ -74,40 +93,58 @@ func (p *planner) Insert(n *parser.Insert) (planNode, error) {
 	primaryIndex := tableDesc.PrimaryIndex
 	primaryIndexKeyPrefix := MakeIndexKeyPrefix(tableDesc.ID, primaryIndex.ID)
 
-	b := client.Batch{}
-
+	var b client.Batch
+	var writes []write
 	for rows.Next() {
-		values := rows.Values()
-		for range cols[len(values):] {
-			values = append(values, parser.DNull)
+		rowVals := rows.Values()
+		for range cols[len(rowVals):] {
+			rowVals = append(rowVals, parser.DNull)
 		}
 
 		for _, col := range tableDesc.Columns {
 			if !col.Nullable {
-				if i, ok := colIDtoRowIndex[col.ID]; !ok || values[i] == parser.DNull {
+				if i, ok := colIDtoRowIndex[col.ID]; !ok || rowVals[i] == parser.DNull {
 					return nil, fmt.Errorf("null value in column %q violates not-null constraint", col.Name)
 				}
 			}
 		}
 
 		primaryIndexKey, _, err := encodeIndexKey(
-			primaryIndex.ColumnIDs, colIDtoRowIndex, values, primaryIndexKeyPrefix)
+			primaryIndex.ColumnIDs, colIDtoRowIndex, rowVals, primaryIndexKeyPrefix)
 		if err != nil {
 			return nil, err
 		}
 
 		// Write the secondary indexes.
 		secondaryIndexEntries, err := encodeSecondaryIndexes(
-			tableDesc.ID, tableDesc.Indexes, colIDtoRowIndex, values)
+			tableDesc.ID, tableDesc.Indexes, colIDtoRowIndex, rowVals)
 		if err != nil {
 			return nil, err
 		}
 
-		for _, secondaryIndexEntry := range secondaryIndexEntries {
+		for i, secondaryIndexEntry := range secondaryIndexEntries {
 			if log.V(2) {
 				log.Infof("CPut %q -> %v", secondaryIndexEntry.key, secondaryIndexEntry.value)
 			}
 			b.CPut(secondaryIndexEntry.key, secondaryIndexEntry.value, nil)
+
+			w := write{
+				values: make(writeMap),
+			}
+
+			indexDesc := &tableDesc.Indexes[i]
+			for _, columnID := range indexDesc.ColumnIDs {
+				colDesc, err := tableDesc.FindColumnByID(columnID)
+				if err != nil {
+					return nil, err
+				}
+
+				w.values[colDesc] = rowVals[colIDtoRowIndex[columnID]]
```

Comment:
Do you really need a map from column descriptor to value? Seems like it would be easier and just as correct to keep a slice of values ordered the same as the index's column ids.